File: test/cases/audit/php/worker.php
<?php
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

declare(strict_types=1);
require_once __DIR__ . '/../../../../lib/php/src/FieldMap.php';
require_once __DIR__ . '/../../../../lib/php/src/Worker.php';
require_once __DIR__ . '/../../../../lib/php/src/Attributes/SerifyModel.php';
require_once __DIR__ . '/../../../../lib/php/src/Attributes/SerifyField.php';
require_once __DIR__ . '/../../../../lib/php/src/SerifyModelHelper.php';
use Serify\Attributes\SerifyField;
use Serify\Attributes\SerifyModel;
use Serify\FieldMap; use Serify\Type; use Serify\Worker;

$unstableCtr = 0; $deserUnstableCtr = 0;
// audit_model keeps its own counter: sharing one would leave this type
// starting wherever `audit` left off, desyncing it across languages.
$modelUnstableCtr = 0;

function _marshal(FieldMap $fm): string {
    $val = $fm->getU32('value'); $tag = $fm->getString('tag');
    $payload = $fm->getBytes('payload'); $tags = $fm->getListString('tags');
    $out = pack('V', $val) . chr(strlen($tag)) . $tag;
    $out .= pack('V', strlen($payload)) . $payload;
    $out .= chr(count($tags));
    foreach ($tags as $t) { $out .= chr(strlen($t)) . $t; }
    return $out;
}
function _unmarshal(string $data, bool $copy=true): FieldMap {
    $fm = new FieldMap(); $off = 0;
    $fm->setU32('value', unpack('V', substr($data, $off, 4))[1]); $off += 4;
    $tlen = ord($data[$off++]); $fm->setString('tag', substr($data, $off, $tlen)); $off += $tlen;
    $plen = unpack('V', substr($data, $off, 4))[1]; $off += 4;
    $p = substr($data, $off, $plen); $fm->setBytes('payload', $copy ? $p : $p); $off += $plen;
    $tcount = ord($data[$off++]); $tags = [];
    for ($i=0; $i<$tcount; $i++) { $tl=ord($data[$off++]); $tags[]=substr($data,$off,$tl); $off+=$tl; }
    $fm->setListString('tags', $tags);
    return $fm;
}
function cleanSer(FieldMap $fm): string { return _marshal($fm); }
function cleanDeser(string $d): FieldMap { return _unmarshal($d, true); }
function mutatingSer(FieldMap $fm): string { $b=_marshal($fm); $fm->setU32('value',0); return $b; }
function unstableSer(FieldMap $fm): string { global $unstableCtr; return _marshal($fm).chr($unstableCtr++); }
function deserUnstableDeser(string $d): FieldMap { global $deserUnstableCtr; $fm=_unmarshal($d,true); if($deserUnstableCtr++>0)$fm->setU32('value',$fm->getU32('value')+1); return $fm; }
function inputMutDeser(string $d): FieldMap { $fm=_unmarshal($d,true); if(strlen($d))$d[0]=chr(ord($d[0])^0xFF); return $fm; }

// ── audit_model: the same faults through the model path ─────────────────────

#[SerifyModel]
class AuditModel
{
    #[SerifyField] public string $payload = '';
    #[SerifyField] public string $tag = '';
    #[SerifyField] public int $value = 0;
    #[SerifyField] public array $tags = [];
}

function _marshalModel(AuditModel $m): string {
    $out = pack('V', $m->value) . chr(strlen($m->tag)) . $m->tag;
    $out .= pack('V', strlen($m->payload)) . $m->payload;
    $out .= chr(count($m->tags));
    foreach ($m->tags as $t) { $out .= chr(strlen($t)) . $t; }
    return $out;
}
function _unmarshalModel(string $data): AuditModel {
    $m = new AuditModel(); $off = 0;
    $m->value = unpack('V', substr($data, $off, 4))[1]; $off += 4;
    $tlen = ord($data[$off++]); $m->tag = substr($data, $off, $tlen); $off += $tlen;
    $plen = unpack('V', substr($data, $off, 4))[1]; $off += 4;
    $m->payload = substr($data, $off, $plen); $off += $plen;
    $tcount = ord($data[$off++]); $tags = [];
    for ($i=0; $i<$tcount; $i++) { $tl=ord($data[$off++]); $tags[]=substr($data,$off,$tl); $off+=$tl; }
    $m->tags = $tags;
    return $m;
}
function modelCleanSer(AuditModel $m): string { return _marshalModel($m); }
function modelMutatingSer(AuditModel $m): string { $b=_marshalModel($m); $m->value=0; return $b; }
/** The positive control: this fault shows in the returned bytes, so it reports
 *  whether or not the model survives the call. */
function modelUnstableSer(AuditModel $m): string { global $modelUnstableCtr; return _marshalModel($m).chr($modelUnstableCtr++); }

Worker::runSuite([
  'audit'=>['clean'=>['cleanSer','cleanDeser'],'mutating'=>['mutatingSer','cleanDeser'],'unstable'=>['unstableSer','cleanDeser'],'deser-unstable'=>['cleanSer','deserUnstableDeser'],'input-mutating'=>['cleanSer','inputMutDeser']],
  'audit_model'=>new Type(AuditModel::class, [
    'clean'=>['modelCleanSer','_unmarshalModel'],
    'mutating'=>['modelMutatingSer','_unmarshalModel'],
    'unstable'=>['modelUnstableSer','_unmarshalModel'],
  ]),
]);

File: test/cases/audit/python/worker.py
# Copyright 2026 Chengxi Luo
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Python audit worker: 5 cross-language audit formats."""
import os, struct, sys
_lib_dir = os.path.join(os.path.dirname(__file__), '../../../../lib/python')
if not os.path.isdir(_lib_dir):
    raise RuntimeError(f"serify library not found at {_lib_dir}; fix the relative path")
sys.path.insert(0, _lib_dir)
from serify import FieldMap, run_suite  # noqa: E402

_unstable_ctr = 0
_deser_unstable_ctr = 0


def _marshal(fm):
    val = fm.get_u32("value")
    tag = fm.get_string("tag").encode()
    payload = fm.get_bytes("payload")
    tags = fm.get_list_string("tags")
    buf = struct.pack("<I", val)
    buf += struct.pack("<B", len(tag)) + tag
    buf += struct.pack("<I", len(payload)) + payload
    buf += struct.pack("<B", len(tags))
    for t in tags:
        b = t.encode()
        buf += struct.pack("<B", len(b)) + b
    return buf


def _unmarshal(data, copy_payload=True):
    fm = FieldMap()
    off = 0
    fm.set_u32("value", struct.unpack_from("<I", data, off)[0]); off += 4
    tlen = data[off]; off += 1
    fm.set_string("tag", data[off:off + tlen].decode()); off += tlen
    plen = struct.unpack_from("<I", data, off)[0]; off += 4
    payload = data[off:off + plen]; off += plen
    fm.set_bytes("payload", bytes(payload) if copy_payload else payload)
    tcount = data[off]; off += 1
    tags = []
    for _ in range(tcount):
        tl = data[off]; off += 1
        tags.append(data[off:off + tl].decode()); off += tl
    fm.set_list_string("tags", tags)
    return fm


def clean_ser(fm):
    return _marshal(fm)


def clean_deser(data):
    return _unmarshal(data, True)


def mutating_ser(fm):
    buf = _marshal(fm)
    fm.set_u32("value", 0)
    return buf


def unstable_ser(fm):
    global _unstable_ctr
    buf = _marshal(fm) + struct.pack("<B", _unstable_ctr)
    _unstable_ctr += 1
    return buf


def deser_unstable_deser(data):
    global _deser_unstable_ctr
    fm = _unmarshal(data, True)
    if _deser_unstable_ctr > 0:
        fm.set_u32("value", fm.get_u32("value") + 1)
    _deser_unstable_ctr += 1
    return fm


def input_mutating_deser(data):
    fm = _unmarshal(data, True)
    ba = bytearray(data)
    if ba:
        ba[0] ^= 0xFF
    return fm


if __name__ == '__main__':
    run_suite({"audit": {
        "clean": (clean_ser, clean_deser),
        "mutating": (mutating_ser, clean_deser),
        "unstable": (unstable_ser, clean_deser),
        "deser-unstable": (clean_ser, deser_unstable_deser),
        "input-mutating": (clean_ser, input_mutating_deser),
    }})
